sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs

```diff
@@ -289,58 +287,66 @@ public async ValueTask<string> GetHeaderValueAsync(HttpMessage message, TokenReq
             private (TaskCompletionSource<HeaderValueInfo> InfoTcs, TaskCompletionSource<HeaderValueInfo>? BackgroundUpdateTcs, bool GetTokenFromCredential)
                 GetTaskCompletionSources(TokenRequestContext context)
             {
+                var localState = _state;
+                // Check if the current state requires no updates to _state under lock and is valid.
+                // All checks must be done on localState as it can't be modified by any other threads.
+                if (localState.InfoTcs != null && !RequestRequiresNewToken(context, localState.CurrentContext) && localState.InfoTcs.Task.IsCompleted && localState.BackgroundUpdateTcs == null)
+                {
+                    DateTimeOffset now = DateTimeOffset.UtcNow;
+                    if (localState.InfoTcs.Task.Status == TaskStatus.RanToCompletion && now < localState.InfoTcs.Task.Result.ExpiresOn && now < localState.InfoTcs.Task.Result.RefreshOn)
+                    {
+                        // localState entity has a valid token, no need to enter lock.
+                        return (localState.InfoTcs, default, false);
+                    }
+                }
                 lock (_syncObj)
                 {
                     // Initial state. GetTaskCompletionSources has been called for the first time
-                    if (_infoTcs == null || RequestRequiresNewToken(context))
+                    if (_state.InfoTcs == null || RequestRequiresNewToken(context, _state.CurrentContext))
                     {
-                        _currentContext = context;
-                        _infoTcs = new TaskCompletionSource<HeaderValueInfo>(TaskCreationOptions.RunContinuationsAsynchronously);
-                        _backgroundUpdateTcs = default;
-                        return (_infoTcs, _backgroundUpdateTcs, true);
+                        _state = new TokenRequestState(context, new TaskCompletionSource<HeaderValueInfo>(TaskCreationOptions.RunContinuationsAsynchronously), default);
+                        return (_state.InfoTcs, _state.BackgroundUpdateTcs, true);
                     }
 
                     // Getting new access token is in progress, wait for it
-                    if (!_infoTcs.Task.IsCompleted)
+                    if (!_state.InfoTcs.Task.IsCompleted)
                     {
-                        _backgroundUpdateTcs = default;
-                        return (_infoTcs, _backgroundUpdateTcs, false);
+                        _state = new TokenRequestState(_state.CurrentContext, _state.InfoTcs, default);
```

Comment:
New instance of `TokenRequestState` is required only if `BackgroundUpdateTcs != null`.